### ISSUE TITLE
Refresh Kubernetes service account tokens automatically

### ISSUE THREAD

**Issue by ypicard** [fluent/fluent-bit#5445]
## Bug Report

**Describe the bug**
I am using fluent-bit (chart 0.20.0, app version 1.9.3) on AWS EKS and received the following email from AWS:

> Kubernetes version 1.21 graduated BoundServiceAccountTokenVolume feature [1] to beta and enabled it by default. This feature improves security of service account tokens by requiring a one hour expiry time, over the previous default of no expiration. This means that applications that do not refetch service account tokens periodically will receive an HTTP 401 unauthorized error response on requests to Kubernetes API server with expired tokens. You can learn more about the BoundServiceAccountToken feature in EKS Kubernetes 1.21 release notes [2].
> To enable a smooth migration of applications to the newer time-bound service account tokens, EKS v1.21+ extends the lifetime of service account tokens to 90 days. Applications on EKS v1.21+ clusters that make API server requests with tokens that are older than 90 days will receive an HTTP 401 unauthorized  error response.
> 
> How can you resolve the issue?
> 
> To make the transition to time bound service account tokens easier, Kubernetes has updated the below official versions of client SDKs to automatically refetch tokens before the one hour expiration:
> 
> * Go v0.15.7 and later
> * Python v12.0.0 and later
> * Java v9.0.0 and later
> * Javascript v0.10.3 and later
> * Ruby master branch
> * Haskell v0.3.0.0
> 
> We recommend that you update your application and its dependencies to use one of the above client SDK versions if you are on an older version.

Apparently, my fluent-bit pod is not refreshing its credentials when using AWS IAM Roles for Service Accounts (IRSA).

**To Reproduce**
- Deploy an AWS EKS cluster with a fluent-bit release (chart 0.20, app version 1.9.3), use AWS IRSA for credentials.

**Expected behavior**
I think the fluent-bit pod should automatically refresh its IRSA credentials.

**Your Environment**

* Version used: 
  * AWS EKS 1.21
  * Fluent bit 1.9.3 (deployed with helm chart 0.20.0)

**Additional context**
I am not 100% sure this needs to be reported here, but I did a few searches in this repository's issues and have not found anything related to this. Let me know if I should investigate somewhere else.

**Comment by sdickhoven** [comment]
what aws is reporting here has nothing to do with irsa (iam for service accounts)... which, granted, also creates secrets with web identity tokens that are mapped into your pods.

but those tokens are used by aws libraries in order to interact with the aws api... and since those web identity tokens have always had a short expiration time, the aws libraries should already be doing the right thing there.

instead, they are talking about the _kubernetes_ service account token that fluentbit uses to interrogate the kube api for metadata.

see: https://docs.fluentbit.io/manual/pipeline/filters/kubernetes

so the kubernetes libraries must be updated (and already have been for various languages besides c) to keep refreshing the service account token before they interact with the kubernetes api.

**Comment by niroowns** [comment]
agreed this is not IRSA related, but what is the take away here for fluentbit though? 

The c libraries need to be updated such that the process actually refreshes the token every hour (or a little before hand), correct?

**Comment by sdickhoven** [comment]
🤔 hmm. looks like fluentbit is doing this itself (no kube client libraries):

https://github.com/fluent/fluent-bit/blob/487a054/plugins/filter_kubernetes/kubernetes.c#L28-L31

but i see logic for refreshing the token every 10 minutes:

https://github.com/fluent/fluent-bit/blob/487a054/plugins/filter_kubernetes/kube_meta.c#L48
https://github.com/fluent/fluent-bit/blob/487a054/plugins/filter_kubernetes/kube_meta.c#L207-L230

so maybe this logic isn't working properly? 🤷

**Comment by jyotimahapatra** [comment]
https://github.com/fluent/fluent-bit/pull/4487 fixes the issue

**Comment by sdickhoven** [comment]
ok. turns out that we're using fluentbit 1.8.1 which was released _before_ the above change.

my bad. 🤦 

we're actually using [newrelic-fluent-bit-output](https://github.com/newrelic/newrelic-fluent-bit-output) v1.10.0 which, in turn, pulls in fluentbit 1.8.1 [here](https://github.com/newrelic/newrelic-fluent-bit-output/blob/v1.10.0/Dockerfile#L20).

so i'll go ahead and upgrade our [newrelic-fluent-bit-output](https://github.com/newrelic/newrelic-fluent-bit-output) to v1.12.3 which pulls in fluentbit 1.8.12.

...of course that doesn't explain why @ypicard is seeing these errors with fluentbit 1.9.3.

however, that could be a red herring because the email we got from aws _today_ said:

>As of April 20th 2022, we have identified the below service accounts attached to pods in one or more of your EKS clusters using stale (older than 1 hour) tokens.

(i.e. this census is from almost a month ago).

so maybe @ypicard _was_ using an older version of fluentbit when aws took that snapshot but has since upgraded to the latest fluentbit version. 🤷 

by the way, @ypicard, if you're shipping kube audit logs to cloudwatch you can check if you are currently getting these warnings with this cloudwatch insights query:

```
fields @timestamp
| filter @logStream like /kube-apiserver-audit-/
| filter @message like /seconds after warning threshold/
| parse @message "subject: *, seconds after warning threshold:*\"" as subject, elapsedtime
```

⚠️ note that [amazon's documentation](https://docs.aws.amazon.com/eks/latest/userguide/service-accounts.html#identify-pods-using-stale-tokens) currently has an errant `i` in their cloudwatch query.

**Comment by sdickhoven** [comment]
sorry for the unnecessary work from my side @jyotimahapatra and thanks for helping out so quickly. 🤗

**Comment by ypicard** [comment]
Hi all,

Thank you for being so responsive here. It is indeed surprising the snapshot was taking a month ago - I had not noticed that. I can confirm @sdickhoven that I have upgraded the chart version to deploy `fluent-bit` last week (from `0.19.19` to `0.20.0`), thus the snapshot is from before.

I ran the above Cloudwatch query and have not no logs were found.

I would consider this issue as solved for now, and will come back to you if I notice anything abnormal in the future.

Thank you!

**Comment by niroowns** [comment]
Sorry guys, I'm not following.  

I just bumped, to test our fluentbit image version from 1.8.5 to 1.8.12, but the issue still persists (the pods need to be up for more than 1 hour for it them to show up in the audit logs, but they eventually get there).  

I also don't follow the thread there, as @jyotimahapatra alluded to: #4487 fixes the issue, but that hasn't made it's way into any formal release yet (PR isn't merged).

**Comment by sdickhoven** [comment]
🤔 hmm. you're right.

i do see code that refreshes the token every 10 minutes but it appears to be meant for a different use case.

https://github.com/fluent/fluent-bit/blob/1fa0e94/plugins/filter_kubernetes/kube_meta.c#L207-L230

specifically, it only runs if `Kube_Token_Command` is configured.

https://github.com/fluent/fluent-bit/blob/1fa0e94/plugins/filter_kubernetes/kube_meta.c#L213

it may be possible to "activate" that code with

```
    [FILTER]
        Name                kubernetes
        Kube_Token_Command  cat /var/run/secrets/kubernetes.io/serviceaccount/token
        ...
```

granted, that's a terrible hack.

the real solution would be to re-read the service account token every 10 minutes (possibly configurable) by default... which is what the above pr wants to do.

**Comment by PettitWesley** [comment]
We cherry-picked the `kube_token_ttl` fix into the latest AWS release: https://github.com/aws/aws-for-fluent-bit/releases/tag/v2.25.0

@sdickhoven

**Comment by PettitWesley** [comment]
@niroowns See above ^ if you use the AWS distro.

**Comment by sdickhoven** [comment]
so...

if you're an impatient person (like me) and you're not using [aws-for-fluent-bit](https://github.com/aws/aws-for-fluent-bit) (like me), i have a workaround to offer:

my [above hack](https://github.com/fluent/fluent-bit/issues/5445#issuecomment-1126256183) to engage the existing token refresh logic in fluentbit totally works.

the only problem is that the fluentbit docker image (and [newrelic's derivative thereof](https://github.com/newrelic/newrelic-fluent-bit-output)) have neither `sh` nor `cat`... which are both required for this to work.

but since kubernetes lets you do wonderful and terrible things, i solved that problem by

1) adding an `emptyDir` volume to the fluentbit pod

```yaml
volumes:
- name: hack
  emptyDir: {}
...
```

2) mounting that under `/bin` using a `volumeMount` in the fluentbit container

```yaml
volumeMounts:
- name: hack
  mountPath: /bin
...
```

3) using the following `initContainer` to populate the volume

```yaml
initContainers:
- name: hack
  image: busybox:1.35
  command:
  - sh
  - -ec
  - |
    cp /bin/busybox /hack/busybox
    ln /hack/busybox /hack/sh
    ln /hack/busybox /hack/cat
  resources:
    limits:
      cpu: 20m
      memory: 128Mi
    requests:
      cpu: 10m
      memory: 64Mi
  volumeMounts:
  - name: hack
    mountPath: /hack
```

i was then able to add the following config to my `fluent-bit.conf`.

```ini
[FILTER]
    Name                kubernetes
    Kube_Token_Command  /bin/cat /var/run/secrets/kubernetes.io/serviceaccount/token
    ...
```

and now i am no longer getting those "seconds after warning threshold" messages in my kube audit log. 🎉 

to be clear... this is a **hack** and not a **solution**! works like a charm, though. 🤗 

(this should work fine in multi-arch clusters too since both `busybox` and `fluent/fluent-bit` are multi-arch images)

⚠️ note: if you are not using `imagePullSecrets` with dockerhub credentials (or some kind of pull through cache) to pull images from dockerhub then this init container will eat into your [6h image pull quota](https://docs.docker.com/docker-hub/download-rate-limit/).

**Comment by GaiheiluKamei** [comment]
Hello, everyone, Please forgive a newbie, here I read all this issue but I am still in confuse.
1. Before I deploy fluent-bit on my EKS with this tutorial: [https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Container-Insights-setup-logs-FluentBit.html](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Container-Insights-setup-logs-FluentBit.html)

2. And after I get Amazon warning, I changed last step's yaml file, just changed the image to: `amazon/aws-for-fluent-bit:2.25.0` and apply the file again, then I exec command `kubectl rollout restart ds/fluent-bit -n amazon-cloudwatch`.

3. So now did i fix this problem, if not then what should I do more?

**Comment by scarecrow111** [comment]
> ## Bug Report
> **Describe the bug** I am using fluent-bit (chart 0.20.0, app version 1.9.3) on AWS EKS and received the following email from AWS:
> 
> > Kubernetes version 1.21 graduated BoundServiceAccountTokenVolume feature [1] to beta and enabled it by default. This feature improves security of service account tokens by requiring a one hour expiry time, over the previous default of no expiration. This means that applications that do not refetch service account tokens periodically will receive an HTTP 401 unauthorized error response on requests to Kubernetes API server with expired tokens. You can learn more about the BoundServiceAccountToken feature in EKS Kubernetes 1.21 release notes [2].
> > To enable a smooth migration of applications to the newer time-bound service account tokens, EKS v1.21+ extends the lifetime of service account tokens to 90 days. Applications on EKS v1.21+ clusters that make API server requests with tokens that are older than 90 days will receive an HTTP 401 unauthorized  error response.
> > How can you resolve the issue?
> > To make the transition to time bound service account tokens easier, Kubernetes has updated the below official versions of client SDKs to automatically refetch tokens before the one hour expiration:
> > 
> > * Go v0.15.7 and later
> > * Python v12.0.0 and later
> > * Java v9.0.0 and later
> > * Javascript v0.10.3 and later
> > * Ruby master branch
> > * Haskell v0.3.0.0
> > 
> > We recommend that you update your application and its dependencies to use one of the above client SDK versions if you are on an older version.
> 
> Apparently, my fluent-bit pod is not refreshing its credentials when using AWS IAM Roles for Service Accounts (IRSA).
> 
> **To Reproduce**
> 
> * Deploy an AWS EKS cluster with a fluent-bit release (chart 0.20, app version 1.9.3), use AWS IRSA for credentials.
> 
> **Expected behavior** I think the fluent-bit pod should automatically refresh its IRSA credentials.
> 
> **Your Environment**
> 
> * Version used:
>   
>   * AWS EKS 1.21
>   * Fluent bit 1.9.3 (deployed with helm chart 0.20.0)
> 
> **Additional context** I am not 100% sure this needs to be reported here, but I did a few searches in this repository's issues and have not found anything related to this. Let me know if I should investigate somewhere else.

Hi, what is the version number of eks you are using?

**Comment by stevehipwell** [comment]
@scarecrow111 this will impact K8s v1.21 & v1.22, so the same for EKS although AWS have limited the impact by allowing tokens to be considered valid for 90 days.

**Comment by prashil-sophos** [comment]
when will the fix be available?

**Comment by scarecrow111** [comment]
The current situation is：
1 My EKS cluster version is 1.21.
2 I also have received the email from AWS official .
3 Found the problem of stale token through the officially provided method
`fields @timestamp
| filter @logStream like /kube-apiserver-audit/
| filter @message like /seconds after warning threshold/
| parse @message "subject: *, seconds after warning threshold:*\"" as subject, elapsedtime `
4 But when I log into the container, I find that the token is updated regularly
Please let me know why, thanks.

**Comment by sdickhoven** [comment]
@scarecrow111 

>But when I log into the container, I find that the token is updated regularly

yes. that's the point.

kubernetes 1.21+ updates the service account token at least once an hour. so when you "log into the container" you will see a new token all the time.

but fluentbit does **not** reread the updated token. it keeps using the token that it read when it first started up.

see: https://github.com/fluent/fluent-bit/blob/v1.9.3/plugins/filter_kubernetes/kube_meta.c#L171-L173

**Comment by sdickhoven** [comment]
@scarecrow111

re 1) see https://github.com/fluent/fluent-bit/issues/5445#issuecomment-1126260881 (yes, you have to upgrade to v2.25+)
re 2) not sure what "fluent client" is. 🤷

**Comment by scarecrow111** [comment]
@sdickhoven Sorry, it should be fluentD client.

**Comment by sdickhoven** [comment]
@scarecrow111 sorry. i don't really know much about fluentd and fluentbit. we ship our logs to newrelic using [this helm chart](https://github.com/newrelic/helm-charts/tree/newrelic-logging-1.10.9/charts/newrelic-logging). the helm chart, in turn, uses [this derivative of fluentbit](https://github.com/newrelic/newrelic-fluent-bit-output). i just started digging through source code when i got those emails from aws and ended up here. i'm not an expert. 🤷

**Comment by scarecrow111** [comment]
@sdickhoven Thank you very much, I have replaced fluentD with fluent-bit-2.10.0, and then check the message through cloudwatch insight, there is no similar stale-token problem, but another cluster also uses fluent-bit-2.10.0, But it prompts the stale-token problem, I am very confused. ref link:https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Container-Insights-setup-EKS-quickstart.html

**Comment by sdickhoven** [comment]
@scarecrow111 please read all the comments above.

your confusion is understandable since both myself and @ypicard fell into this trap.

but @niroowns already pointed out [here](https://github.com/fluent/fluent-bit/issues/5445#issuecomment-1126074994) that:

>the pods need to be up for more than 1 hour for them to show up in the audit logs

**Comment by scarecrow111** [comment]
@sdickhoven Thanks for your help.

**Comment by sdickhoven** [comment]
regarding [this](https://github.com/fluent/fluent-bit/issues/5445#issuecomment-1125547287):

an alternative approach i've seen for refreshing the service account token is to keep using the current token for as long as the api accepts it.

once the api returns a 401, reread the token, compare it to the previous token, if it's different (i.e. if you're **not** getting a 401 for reasons other than token expiration), retry request with new token.

that way you're not refreshing the token on a cadence but do it "on-demand" (and only as often as needed) instead.

**Comment by stevehipwell** [comment]
@sdickhoven if a token has an expiration and it's close or passed it should be refreshed. If a refresh period has been set this should follow similar behaviour.

Your solution is only valid if neither of the above are available. It would lead to 29 days and 23 hours worth of warning logs in EKS, lots of error logs in all clusters, and the potential to regularly hit a slow path on the API server causing degradation of service.

**Comment by sdickhoven** [comment]
>This way you'll get a bunch of 401s in your logs + failed attempts by fluentd/fluent-bit kubernetes metadata plugin to read K8s metadata & as a result of that your logs from k8s missing all that data which is a mess.

that's how the official kubernetes client-go library does it. 🤷 

>It would lead to 29 days and 23 hours worth of warning logs in EKS

yes. indeed.

and do i see these warnings in my kube-audit log for workloads that are using the most recent version of the official kubernetes client-go library.

but those warnings are spurious and e.g. [flux](https://github.com/fluxcd/flux2/blob/main/go.mod#L43) (or whatever other workloads use the client-go library) will not actually stop working when 90d (not 30d) have elapsed.

**Comment by lecaros** [comment]
fix shipped with [1.9.4](https://github.com/fluent/fluent-bit/releases/tag/v1.9.4).
